Initial alpha release of Fiber Connectivity Viewer

martcous opened the pull request:
- New "Fibers" data type
- Support of importing TRK files
- 3D Viewer can display fiber streamlines with surface objects
- Displayed fibers are colored with gradients based on their curvature and orientation
- Connectivity graphs can be displayed with associated fibers
- Fibers are associated to scouts with simplistic method of minimum distance of endpoints to scout centroid. To be improved. An [experimental] label is added to warn users that this is not final.
- Can toggle between connections that are "anatomically accurate" a.k.a. have an associated fibers vs connections that do not.

![image](https://user-images.githubusercontent.com/9253273/55568092-07c2c480-56cd-11e9-9043-261bf9ba16c2.png)
![image](https://user-images.githubusercontent.com/9253273/55568433-b1a25100-56cd-11e9-8d67-d6f5d8a59304.png)


What's next:
- Uploading a template TRK file to neuroimage.usc.edu. Here's an example with 3000 fibers: https://www.dropbox.com/s/gld98uhtbtjyem4/whole_brain_MNI_3k.trk?dl=0
- Adding a download menu in software for fiber template once uploaded
- Making a tutorial page once PR is approved. Need to cite the following paper for using fiber atlas: https://www.ncbi.nlm.nih.gov/pubmed/29758339

martcous commented on external/trk/trk_interp.m:
```diff
@@ -0,0 +1,143 @@
+%    [header_sc tracks_sc]   = trk_add_sc(header, tracks_interp_str, volume, 'FA');
+%    [scalar_mean scalar_sd] = trk_mean_sc(header_sc, tracks_sc);
+%
+% Other m-files required: Curve Fitting Toolbox (aka Spline Toolbox)
```
The Curve Fitting toolbox is only optional so we're ok!

martcous commented on toolbox/gui/figure_3d.m:
```diff
@@ -4248,5 +4349,60 @@ function JumpMaximum(hFig)
+    
+    %% If fibers not yet assigned to atlas, do so now
+    if isempty(FibMat.Scouts(1).ConnectFile) || ~ismember(TfInfo.FileName, {FibMat.Scouts.ConnectFile})
+        ScoutNames     = getappdata(hFigConn, 'RowNames');
```
Awesome, I didn't know about GetFigureHandles/SetFigureHandles! I will refactor that in. Sadly, getappdata/setappdata is used a lot in figure_connect() already... I'll see if I can refactor that as well without breaking anything.

martcous replied:
Actually using the actual functions GetFigureHandles/SetFigureHandles everywhere is not the way to go because they create a local copy of the global variable when not all fields are actually required. I'll instead use GlobalData.DataSet.Figure.Handles directly. Let me know if I'm missing something.

martcous commented on toolbox/gui/figure_3d.m:
```diff
@@ -4248,5 +4349,60 @@ function JumpMaximum(hFig)
+    if isempty(FibMat.Scouts(1).ConnectFile) || ~ismember(TfInfo.FileName, {FibMat.Scouts.ConnectFile})
+        ScoutNames     = getappdata(hFigConn, 'RowNames');
+        ScoutCentroids = getappdata(hFigConn, 'RowLocs');
+        FibMat = import_fibers('AssignToScouts', FibMat, TfInfo.FileName, ScoutCentroids);
```
Sure, I don't mind moving this function elsewhere. I was using import_fibers as a "fibers util" kind of file. Moving it to figure_connect would make sense I think, you're right.

martcous commented on toolbox/io/in_fibers.m:
```diff
@@ -0,0 +1,130 @@
+    end
+end
+        
+%% ===== CONVERSION MRI TO SCS =====
```
I am not sure why this would be needed. Fibers file are always in MRI coordinates as far as I know?

ftadel replied:
The templates will probably be in MNI coordinates, and if you want to load them for a given subject, you might need something like this. 
"MRI coordinates" mean different things in each program. Actually, many programs would rather use "world coordinates" to add 3D coordinates on top of a volume: https://neuroimage.usc.edu/brainstorm/CoordinateSystems#World_coordinates

This is a cleaner way to proceed than what is done in the function ApplyMriTransf(). I understood this only 3-4 years ago, I cleaned a bit the code accordingly (with cs_convert/cs_compute), but some functions still have this old logic (like the in_tess function you copied).

But maybe you can wait until someone complains that their own models are not aligned with the volumes and surfaces in Brainstorm, to see what they need exactly before writing useless code none would one ever use :)

martcous replied:
Let's wait until someone complains :)

martcous commented on toolbox/io/import_fibers.m:
```diff
@@ -0,0 +1,353 @@
+%  ===== HELPER FUNCTIONS ===============================================================
+%  ======================================================================================
+%% ===== APPLY MRI ORIENTATION =====
+function sSurf = ApplyMriTransf(MriTransf, sSurf)
```
I did need this in some cases to align the fibers with the MRI. It's not about converting the coordinates, just to apply a translation.

martcous commented on toolbox/gui/view_connect.m:
```diff
@@ -172,12 +181,56 @@
 setappdata(hFig, 'DataFile',     GlobalData.DataSet(iDS).DataFile);
 setappdata(hFig, 'StudyFile',    GlobalData.DataSet(iDS).StudyFile);
 setappdata(hFig, 'SubjectFile',  GlobalData.DataSet(iDS).SubjectFile);
+setappdata(hFig, 'plotFibers',   plotFibers);
```
I kept these ones because they were such small size variables (I couldn't move everything into Handles because some were set before the figure was registered in Brainstorm).

martcous commented:
So I moved all big variables from appdata to GlobalData handles and WOW the graph connectivity viewer is now much smoother for me. Thanks a lot for the suggestion, this will greatly benefit all our users!

I addressed all your comments. You can test further using the fiber file attached above.
1. Generate a connectivity file on a subject with the ICMB152 template set as the anatomy.
2. You can then import the fiber file from the anatomy viewer (right click on the subject -> Import fibers). 
3. Then, you can either double click on the new object to display the fibers or go to your connectivity file and right click -> Display fibers

All features are listed in the PR description. Let me know if you have more questions!

ftadel commented:
> Uploading a template TRK file to neuroimage.usc.edu:

Done:
https://neuroimage.usc.edu/bst/getupdate.php?d=whole_brain_MNI_3k.trk

ftadel commented:
Todo: 
- Prepare a tutorial with an example dataset.
- Add a summary of the new tools in the News page

ftadel commented:
Another thing that would be great:  rewriting figure_connect without java3d/jogl. 
Now it is possible to produce these graphs only with Matlab objects we don't need this anymore...
The support for JOGL may disappear at some point, and even now it's a cause for crashes on many Linux systems.

ftadel commented:
Just spotted another major problem:
trk_interp uses parfor as it's supposed to be available everywhere...

Most people do not have access to the parallel processing toolbox. Plus its implementation has been changing every 3-4 years (check for the block at line 83 of bst_process).
You need to write an alternative solution for when the toolbox is not accessible, or when we don't want to use it (because it crashes for instance), like it is done for all the code that uses parfor loops (search for "parfor" or "parpool" in the brainstorm code).

You may also consider simply changing parfor with simple for loops... Check if the gain in time is really worth going through the pain of really using parfor loops. If it is for importing something only a few times per protocol, we don't really care if it takes 20s instead of 2s.
The maintenance of parfor loops have been painful to deal with in the past, and I'm sure we're not at the end of it.

ftadel commented:
Except for this, it seems to be working beautifully :-)

ftadel commented:
Let me take this last comment back...
I didn't catch the fact you were actually altering the database structure...
This is unfortunately not possible easily.
Take an existing protocol and try adding a new subject, see what happens.

Please rollback or fix urgently.

ftadel commented:
I commited this before people damage permanently their existing databases: 
https://github.com/brainstorm-tools/brainstorm3/commit/b2404f27c4a3c1b1f5dd24f0d81ecede5720f7ee

I'm sorry about that... I should have seen this when I was reading your code.
This stupid database structure can't be extended, this is 1) why connectivity objects are stored as time-frequency files and 2) why we need a new database engine...

You need to find a way to make this fit in the existing structures...
